src/Transactions.ts

```diff
@@ -0,0 +1,834 @@
+          break;
+        case TransactionStatus.STATUS_COMPLETED:
+          this.sendACK();
+          break;
+      }
+    }
+  }
+
+  public sendACK(options: any = {}): OutgoingRequestType | undefined {
+    // TODO: Move PRACK stuff into the transaction layer. That is really where it should be
+    let ruri: string | URI;
+    if (this.response && this.response.getHeader("contact")) {
+      ruri = this.response.parseHeader("contact").uri;
+    } else {
+      ruri = this.request.ruri;
+    }
+
+    if (this.response) {
+      const ack: OutgoingRequestType = new OutgoingRequest(
+        "ACK",
+        ruri.toString(),
+        this.request.ua,
+        {
+          cseq: this.response.cseq,
+          callId: this.response.callId,
+          fromUri: this.response.from.uri,
+          fromTag: this.response.fromTag,
+          toUri: this.response.to.uri,
+          toTag: this.response.toTag,
+          routeSet: this.response.getHeaders("record-route").reverse()
+        },
+        options.extraHeaders || [],
+        options.body
+      );
+
+      if (!ack.ua.transport) {
+        throw new Error("No transport to make transaction");
+      }
+      this.ackSender = new AckClientTransaction({
+        onTransportError: this.requestSender.applicant ?
+        this.requestSender.applicant.onTransportError :
+        () => {
+          this.logger.warn("ACK Request had a transport error");
+        },
+        ua: ack.ua
+      }, ack, ack.ua.transport);
+      this.ackSender.send();
+
+      return ack;
+    }
+  }
+
+  private onTransportError(): void {
+    this.logger.log("transport error occurred, deleting INVITE client transaction " + this.id);
+    if (this.B) {
+      clearTimeout(this.B);
+      this.B = undefined;
+    }
+    if (this.D) {
+      clearTimeout(this.D);
+      this.D = undefined;
+    }
+    if (this.M) {
+      clearTimeout(this.M);
+      this.M = undefined;
+    }
+
+    this.stateChanged(TransactionStatus.STATUS_TERMINATED);
+    this.requestSender.ua.destroyTransaction(this);
+
+    if (this.state !== TransactionStatus.STATUS_ACCEPTED) {
+      this.requestSender.onTransportError();
+    }
+  }
+
+  // RFC 6026 7.2
+  private timer_M(): void {
+    this.logger.debug("Timer M expired for INVITE client transaction " + this.id);
+
+    if (this.state === TransactionStatus.STATUS_ACCEPTED) {
+      if (this.B) {
+        clearTimeout(this.B);
+        this.B = undefined;
+      }
+      this.stateChanged(TransactionStatus.STATUS_TERMINATED);
+      this.requestSender.ua.destroyTransaction(this);
+    }
+  }
+
+  // RFC 3261 17.1.1
+  private timer_B(): void {
+    this.logger.debug("Timer B expired for INVITE client transaction " + this.id);
+    if (this.state === TransactionStatus.STATUS_CALLING) {
+      this.stateChanged(TransactionStatus.STATUS_TERMINATED);
+      this.requestSender.ua.destroyTransaction(this);
+      this.requestSender.onRequestTimeout();
+    }
+  }
+
+  private timer_D(): void {
```

Comment:
This is never called. Could be a bug or dead code, haven't looked that closely.

Reply:
Yea, one thing I didn't turn on was noUnunusedLocalReferences (something to that effect). I did that for the least amount of changes possible. You'll probably see a good amount of that, I figure it can be cleaned up in a second pass, with fairly low chance of breaking anything (as its all unused).

Reply:
I accidentally removed this timer in an earlier commit (https://github.com/onsip/SIP.js/commit/55cb0aeff36d314e20292dbf8fb1ddbc0f3cd7b0#diff-ef9ade42e4a5eaa513b3d74c3f4f8c0cL249) by accident. It should be added back. I would like to do a full rework on the Transaction code in the future if I can find some time for it.